Documentation/postmortems/v3.5-data-inconsistency.md

```diff
@@ -0,0 +1,121 @@
+Implementation was as follows, before applying the WAL entries, etcd updated in memory value of consistent index. 
+As part of transaction commit process, a database hook would read the value of consistent index and store it to database. 
+
+Problem is that in memory value of consistent index is shared, and there might be other in flight transactions apart from serial WAL apply flow.
+So if we imagine scenario:
+1. Etcd server starts an apply workflow, and it just sets a new consistent index value.
+2. The periodic commit is triggered, and it executes the backend hook and saves consistent index from apply workflow.
+3. Etcd server finished an apply workflow, saves new changes and saves same value of consistent index again.
+
+Between second and third point there is a very small window where consistent index is increased without applying any entries from WAL.
+
+## Trigger
+
+If etcd crashed after consistency index is saved, but before to apply workflow finished it would lead to data inconsistency.
+When recovering the data etcd would skip executing changes from failed apply workflow, assuming they have been already executed.
+
+This follows the issue reports and code used to reproduce the issue where trigger was etcd crashing under high request load.
+Crashes in reports where caused by etcd running under high memory pressure, causing it to go out of memory from time to time.
+Reproduction run etcd under high stress and randomly killed one of the members using SIGKILL signal (not recoverable immediate process death). 
+
+## Detection
+
+For single member cluster it is totally undetectable. 
+There is no mechanism or tool for verifying that state database matches WAL.  
+
+In cluster with multiple members it would mean that one of the members that crashed, will missing changes from failed apply workflow.
+This means that it will have different state of database and will return different hash via `HashKV` grpc call.
+
+There is an automatic mechanism to detect data inconsistency. 
+It can be executed during etcd start via `--experimental-initial-corrupt-check` and periodically via `--experimental-corrupt-check-time`.
+Both checks however have a flaw, they depend on `HashKV` grpc method, which might fail causing the check to pass.
+
+In multi member etcd cluster, each member can run with different performance and be at different stage of applying the WAL log.
+Comparing database hashes between multiple etcd members requires all hashes to be calculated at the same change.
+This is done by requesting hash for the same `revision` (version of key value store). 
+However, it will not work if the provided revision is not available on the members.
+This can happen on very slow members, or in cases where corruption has lead revision numbers to diverge.
+
+This means that for this issue, the corrupt check is only reliable during etcd start just after etcd the crash.
```

Comment:
suggest to remove `the` in `etcd the crash.`

Reply:
Done